From fa519d725b647163fcb61de98c3e354b11b5a803 Mon Sep 17 00:00:00 2001
From: Terry Zhou <bjzhou@marvell.com>
Date: Tue, 15 May 2018 11:10:50 +0800
Subject: [PATCH 0653/1051] arm64: dts: marvell: 7040: add configs for setups B
 C D E

7040 DB board can have different setups for different kinds of
usage by changing the DPR and switches on board.
The detailed differences of these setups can be found in:
Documentation/mvebu/a7k-a8k/armada7040-db-setup.txt
The u3d, udc driver and TDM driver are still missing in the current
kernel source tree, should add dts support after adding these drivers

Change-Id: Iee6f5110c705d55603904f53acc8e29e4d107d38
Signed-off-by: Terry Zhou <bjzhou@marvell.com>
[Kevin: The original patch got from Marvell sdk10.0_19.06]
Signed-off-by: Kevin Hao <kexin.hao@windriver.com>
---
 .../bindings/arm/marvell/armada-7k-8k.txt     |  12 ++
 .../mvebu/a7k-a8k/armada7040-db-setup.txt     | 153 ++++++++++++++++++
 arch/arm64/boot/dts/marvell/Makefile          |   4 +
 .../boot/dts/marvell/armada-7040-db-B.dts     |  80 +++++++++
 .../boot/dts/marvell/armada-7040-db-C.dts     | 107 ++++++++++++
 .../boot/dts/marvell/armada-7040-db-D.dts     | 126 +++++++++++++++
 .../boot/dts/marvell/armada-7040-db-E.dts     | 100 ++++++++++++
 .../boot/dts/marvell/armada-7040-db.dtsi      | 138 ++++++++++++++++
 8 files changed, 720 insertions(+)
 create mode 100644 Documentation/mvebu/a7k-a8k/armada7040-db-setup.txt
 create mode 100644 arch/arm64/boot/dts/marvell/armada-7040-db-B.dts
 create mode 100644 arch/arm64/boot/dts/marvell/armada-7040-db-C.dts
 create mode 100644 arch/arm64/boot/dts/marvell/armada-7040-db-D.dts
 create mode 100644 arch/arm64/boot/dts/marvell/armada-7040-db-E.dts
 create mode 100644 arch/arm64/boot/dts/marvell/armada-7040-db.dtsi

diff --git a/Documentation/devicetree/bindings/arm/marvell/armada-7k-8k.txt b/Documentation/devicetree/bindings/arm/marvell/armada-7k-8k.txt
index 1daadaccf8e2..a516c4ff3322 100644
--- a/Documentation/devicetree/bindings/arm/marvell/armada-7k-8k.txt
+++ b/Documentation/devicetree/bindings/arm/marvell/armada-7k-8k.txt
@@ -22,6 +22,18 @@ Boards:
    - "marvell,armada7040-db-default", "marvell,armada7040-db"
       when the board being used is the Armada 7040 Development board
 
+   - "marvell,armada7040-db-B", "marvell,armada7040-db"
+      when the board being used is the Armada 7040 Development board with Setup B
+
+   - "marvell,armada7040-db-C", "marvell,armada7040-db"
+      when the board being used is the Armada 7040 Development board with Setup C
+
+   - "marvell,armada7040-db-D", "marvell,armada7040-db"
+      when the board being used is the Armada 7040 Development board with Setup D
+
+   - "marvell,armada7040-db-E", "marvell,armada7040-db"
+      when the board being used is the Armada 7040 Development board with Setup E
+
    - "marvell,armada8040-db-default", "marvell,armada8040-db"
       when the board being used is the Armada 8040 Development board
 
diff --git a/Documentation/mvebu/a7k-a8k/armada7040-db-setup.txt b/Documentation/mvebu/a7k-a8k/armada7040-db-setup.txt
new file mode 100644
index 000000000000..71f5fc647248
--- /dev/null
+++ b/Documentation/mvebu/a7k-a8k/armada7040-db-setup.txt
@@ -0,0 +1,153 @@
+Armada7040 development board setup:
+--------------------------------------
+
+the Armada 7040 development board (PCB: DB-88F7040-Modular) is a configurable board,
+designed to test several combinations of SoC configurations.
+linux supports a defined set of those configurations via different device tree files.
+
+This document describes the board modifications required to set up each configuration and the
+interfaces supported (or disabled in each one).
+
+To switch board configuration:
+1. Modify HW board settings according to the instructions in the "Board Setup" section below.
+2. Use appropriate Device Tree file for booting Linux.
+3. Same Device Tree selection is required also in U-Boot (to initialize requested comphy SerDes lanes).
+   Modify the Device Tree file used by U-BOOT, by running the U-BOOT command "fdt_config",
+   and select the appropriate configuration entry.
+
+
+Board Setup
+------------
+For A7040 A revision SoC - Please use armada-7040-db configuration,
+
+|		| 	1: default (A)	|	2:B		|	3:C		|       4:D             |       4:E             |
+|---------------|-----------------------|-----------------------|-----------------------|-----------------------|-----------------------|
+|Device tree	|armada-7040-db.dtb	|armada-7040-db-B.dtb	|armada-7040-db-C.dtb	|armada-7040-db-D.dtb   |armada-7040-db-E.dtb	|
+|CP0-SW1:1	| ON			| OFF			| ON			| ON 			| OFF                   |
+|CP0-SW1:2	| OFF			| ON			| OFF			| OFF 			| OFF                   |
+|CP0-SW1:3	| ON			| ON			| ON			| ON			| ON                    |
+|CP0-SW1:4	| OFF			| ON			| OFF			| OFF			| OFF                   |
+|CP0-SW1:5	| OFF			| ON			| OFF			| OFF			| OFF                   |
+|CP0-SW1:6	| ON			| ON			| ON			| ON			| ON                    |
+|CP0-SW1:7	| ON			| ON			| ON			| ON			| ON                    |
+|CP0-SW1:8	| OFF			| OFF			| OFF			| OFF			| OFF                   |
+----------------------------------------------------------------------------------------------------------------------------------------|
+
+DPR Setting
+------------
+|					| 	1: default (A)			 |	2: B:                             |
+|---------------------------------------|----------------------------------------|----------------------------------------|
+| Mpp Jumper Module (SLM1548)		| Default release (ALL DPRs 1-2 position)| Default release (ALL DPRs 1-2 position)|
+| Serdess Jumper Module (SLM1547)	| DPR152,153,154,155 - 1-2 position	 | DPR152,153,154,155 - 1-2 position      |
+|					| DPR148,149,150,151 - 2-3 position	 | DPR148,149,150,151 - 1-2 position      |
+| CP0 Serdess module (SLM1521)		| SW1 - (1,3,6,7)= ON,  (2,4,5,8)=OFF	 | SW1 - (2,3,4,5,6,7)= ON,  (1,8)=OFF    |
+| 					| DPR2,3,4,5 - 1-2 position		 | DPR2,3,4,5 - 2-3 position              |
+----------------------------------------|----------------------------------------|----------------------------------------|
+|					| 	3: C				 |	4:D					       |	5:E                                          |
+|---------------------------------------|----------------------------------------|-----------------------------------------------------|-----------------------------------------------------|
+| Mpp Jumper Module (SLM1548)		| DPR133 - 143  2-3 position		 | DPR138 - 143  2-3 position,DPR133-137 not connect   | DPR133 -143  1-2 position , DPR144-147 2-3 position |
+| Serdess Jumper Module (SLM1547)	| DPR152,153,154,155 - 1-2 position	 | DPR152,153,154,155 - 1-2 position                   | DPR152,153,154,155 - 1-2 position                   |
+|					| DPR148,149,150,151 - 1-2 position	 | DPR148,149,150,151 - 1-2 position		       | DPR148,149,150,151 - 1-2 position                   |
+| CP0 Serdess module (SLM1521)		| SW1 - (1,3,6,7)= ON,  (2,4,5,8)=OFF	 | SW1 - (1,3,6,7)= ON,  (1,2,4,5,8)=OFF               | SW1 - (3,6,7)= ON,  (1,2,4,5,8)=OFF                 |
+| 					| DPR2,3,4,5 - 1-2 position		 | DPR2,3,4,5 - 1-2 position                           | DPR2,3,4,5 - 1-2 position                           |
+|					|					 | CON9 not connect , CON10 connect,                   |                                                     |
+|					|					 | R77 R79 R81-82 R84-85 connect                       |                                                     |
+| Install TDM Module (SLM1448)		| on DB-7040 CON46			 | -                                                   | -                                                   |
+|					|					 |                                                     |                                                     |
+| Install Audio Module (SLM1449)	| - 					 | on DB-7040 CON46.                                   | -                                                   |
+|					|					 |                                                     |                                                     |
+|					|					 |                                                     |                                                     |
+| DB-4070 Modular (SLM1546)		| - 					 | wire from U36-1 to con47-2, DPR23 -NC               | DPR53-54 1-2 position , DPR61 2-3 position          |
+|					| 					 |                                                     | DPR10-11 position 1-2, DPR12-15 position 1-2        |
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+
+The tables below summarizes the interface configuration of each setup
+
+SerDes Configuration
+------------------------
+
+| CP0 Lane 	| 	1: default (A)	|   2: B	| 3: C            | 4:D		  | 5:E		  |
+|---------------|-----------------------|---------------|-----------------|---------------|---------------|
+| 0		| SGMII1		| PICE0 (x4)	| PCIE0 (x1)      |  PCIE0 (x1)   | PCIE0 (x1)	  |
+| 1		| USB3_HOST0		| PCIE0 (x4)	| SATA0		  |  USB3_DEVICE0 | SATA0	  |
+| 2		| SFI (10G)		| PCIE0 (x4)	| SFI (10G)       |  SFI (1G)  	  | SGMII0 (1G)   |
+| 3		| SATA1			| PCIE0 (x4)	| SATA1		  |  SATA1        | SATA1	  |
+| 4		| USB3_HOST1		| USB3_HOST1	| USB3_HOST1      |  USB3_HOST1   | USB_HOST1	  |
+| 5		| PCIE2 (x1)		| PCIE2 (x1)	| PCIE2 (x1)      |  PCIE2 (x1)   | PCIE2 (x1)	  |
+-----------------------------------------------------------------------------------------------------------
+
+- USB2_H0 (UTMI only)
+- USB2_H1 (UTMI & USB3 SerDes)
+
+
+Multi-purpose pin configurations
+--------------------------------
+
+Setup  #	|	1	|	2	|       3       |	4	|	5	|
+Name   #	|  default (A) 	|	B	|       C       |	D	|	E	|
+-------------------------------------------------------------------------------------------------
+AP806:				|		|		|		|		|
+-------------------------------------------------------------------------------------------------
+	AP-SDIO | [0-5]		| [0-5]		| [0-5]         | [0-5]		| N/C		|
+	AP-SPI0 | N/C		| N/C		| N/C		| N/C		| [0-3]		|
+	AP-I2C	| N/C		| N/C		| N/C		| N/C		| [4,5]		|
+	AP-UART0| [11,19]	| [11,19]	| [11,19]       | [11,19]	| [11,19]	|
+-------------------------------------------------------------------------------------------------
+CP:				|		|		|		|		|
+-------------------------------------------------------------------------------------------------
+	CP-SPI0	| N/C		| N/C		| [6-11]	| N/C		| N/C		|
+	CP-SPI1	| [13-16]	| [13-16]	| [13-16]	| N/C		| [13-16]	|
+	NAND	| N/C		| N/C		| N/C		| [15-27]	| N/C		|
+	RGMII0	| N/C		| [44-55]	| [0-11]	| N/C		| [0-11]	|
+	RGMII1	| [0-11]	| [0-11]	| [44-55]	| [44-55]	| N/C		|
+	CP-UART0| [29-30]	| [29-30]	| [29-30	| [29-30]	| [29-30|	|
+	CP-UART1| N/C		| N/C		| N/C		| N/C		| [47,49-51]	|
+	SD	| [56-61]	| [56-61]	| [56-61]	| [56-61]	| [54,56-61]	|
+	GPIO	| [62]		| [62]		| [62]		| [62]		| [62]		|
+	TDM	| N/C		| N/C		| [0-5]		| N/C		| N/C		|
+	TWSI	| [37-38]	| [37-38]	| [37-38]	| [37-38]	| [37-38]	|
+	SATA0	| N/C		| N/C		| N/C		| N/C		| N/C		|
+	SATA1	| [28]		| [28]		| [28]		| [28]		| [28]		|
+	USB_DEV | N/C 		| N/C 		| N/C		| [7]		| N/C		|
+-------------------------------------------------------------------------------------------------
+
+Network configuration
+---------------------
+1: default (A):
+ Interface	| CP#	| PPv2 Port	| GMAC Port	| Board Interface		|
+ ---------------|-------|---------------|---------------|-------------------------------|
+ eth0		| CP0	| eth0		| 0		| SFI (10G)			|
+ eth1		| CP0	| eth1		| 2		| SGMII1			|
+ eth2		| CP0	| eth2		| 3		| RGMII1			|
+ ----------------------------------------------------------------------------------------
+
+2: B:
+ Interface      | CP#   | PPv2 Port     | GMAC Port     | Board Interface               |
+ ---------------|-------|---------------|---------------|-------------------------------|
+ eth0           | CP0   | eth0          | 2             | RGMII0                        |
+ eth1           | CP0   | eth1          | 3             | RGMII1                        |
+ ----------------------------------------------------------------------------------------
+
+3: C:
+ Interface      | CP#   | PPv2 Port     | GMAC Port     | Board Interface               |
+ ---------------|-------|---------------|---------------|-------------------------------|
+ eth0           | CP0   | eth0          | 0             | SFI (10G)                     |
+ eth1           | CP0   | eth1          | 3             | RGMII1                        |
+ ----------------------------------------------------------------------------------------
+4: D:
+ Interface      | CP#   | PPv2 Port     | GMAC Port     | Board Interface               |
+ ---------------|-------|---------------|---------------|-------------------------------|
+ eth0           | CP0   | eth0          | 0             | SFI (10G)                     |
+ eth1           | CP0   | eth1          | 3             | RGMII1                        |
+ ----------------------------------------------------------------------------------------
+
+5: E:
+ Interface      | CP#   | PPv2 Port     | GMAC Port     | Board Interface               |
+ ---------------|-------|---------------|---------------|-------------------------------|
+ eth0           | CP0   | eth0          | 0             | SGMII0 (1G)                   |
+ eth1           | CP0   | eth1          | 2             | RGMII0                        |
+ ----------------------------------------------------------------------------------------
+
+I2C IO expander0 GPIO pin configuration
+--------------------------------------
+PIN[15] - GPIO regulator of cp0-vccq-sd0
diff --git a/arch/arm64/boot/dts/marvell/Makefile b/arch/arm64/boot/dts/marvell/Makefile
index 4f1dbcbd366c..9bb489fd6648 100644
--- a/arch/arm64/boot/dts/marvell/Makefile
+++ b/arch/arm64/boot/dts/marvell/Makefile
@@ -3,6 +3,10 @@
 dtb-$(CONFIG_ARCH_MVEBU) += armada-3720-db.dtb
 dtb-$(CONFIG_ARCH_MVEBU) += armada-3720-espressobin.dtb
 dtb-$(CONFIG_ARCH_MVEBU) += armada-7040-db.dtb
+dtb-$(CONFIG_ARCH_MVEBU) += armada-7040-db-B.dtb
+dtb-$(CONFIG_ARCH_MVEBU) += armada-7040-db-C.dtb
+dtb-$(CONFIG_ARCH_MVEBU) += armada-7040-db-D.dtb
+dtb-$(CONFIG_ARCH_MVEBU) += armada-7040-db-E.dtb
 dtb-$(CONFIG_ARCH_MVEBU) += armada-8040-db.dtb
 dtb-$(CONFIG_ARCH_MVEBU) += armada-8040-mcbin.dtb
 dtb-$(CONFIG_ARCH_MVEBU) += armada-8040-mcbin-single-shot.dtb
diff --git a/arch/arm64/boot/dts/marvell/armada-7040-db-B.dts b/arch/arm64/boot/dts/marvell/armada-7040-db-B.dts
new file mode 100644
index 000000000000..2929c23ad8e2
--- /dev/null
+++ b/arch/arm64/boot/dts/marvell/armada-7040-db-B.dts
@@ -0,0 +1,80 @@
+// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
+/*
+ * Device Tree file for Marvell Armada 7040 development board
+ * This board file supports the B configuration of the board
+ */
+
+#include "armada-7040-db.dtsi"
+
+/ {
+	model = "Marvell Armada 7040 development board B setup";
+	compatible = "marvell,armada7040-db-B", "marvell,armada7040-db",
+		     "marvell,armada7040", "marvell,armada-ap806-quad",
+		     "marvell,armada-ap806";
+
+	chosen {
+		stdout-path = "serial0:115200n8";
+	};
+
+	aliases {
+		ethernet0 = &cp0_eth1;
+		ethernet1 = &cp0_eth2;
+	};
+};
+
+&ap_sdhci0 {
+	status = "okay";
+};
+
+&cp0_i2c0 {
+	status = "okay";
+	clock-frequency = <100000>;
+};
+
+&cp0_spi1 {
+	status = "okay";
+};
+
+&cp0_pcie0 {
+	status = "okay";
+	num-lanes = <4>;
+	phys = <&cp0_comphy0 0>;
+	phy-names = "pcie-phy0";
+};
+
+&cp0_pcie2 {
+	status = "okay";
+	phys = <&cp0_comphy5 2>;
+	phy-names = "pcie-phy0";
+};
+
+&cp0_usb3_1 {
+	status = "okay";
+};
+
+&cp0_mdio {
+	status = "okay";
+
+	cp0_phy2: ethernet-phy@2 {
+		reg = <0>;
+	};
+	cp0_phy3: ethernet-phy@3 {
+		reg = <1>;
+	};
+};
+
+&cp0_ethernet {
+	status = "okay";
+};
+
+&cp0_eth1 {
+	status = "okay";
+	phy = <&cp0_phy2>;
+	phy-mode = "rgmii-id";
+};
+
+&cp0_eth2 {
+	status = "okay";
+	phy = <&cp0_phy3>;
+	phy-mode = "rgmii-id";
+};
diff --git a/arch/arm64/boot/dts/marvell/armada-7040-db-C.dts b/arch/arm64/boot/dts/marvell/armada-7040-db-C.dts
new file mode 100644
index 000000000000..8ca4a8ac32a7
--- /dev/null
+++ b/arch/arm64/boot/dts/marvell/armada-7040-db-C.dts
@@ -0,0 +1,107 @@
+// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
+/*
+ * Device Tree file for Marvell Armada 7040 development board
+ * This board file supports the C configuration of the board
+ */
+
+#include "armada-7040-db.dtsi"
+
+/ {
+	model = "Marvell Armada 7040 development board C setup";
+	compatible = "marvell,armada7040-db-C", "marvell,armada7040-db",
+		     "marvell,armada7040", "marvell,armada-ap806-quad",
+		     "marvell,armada-ap806";
+
+	chosen {
+		stdout-path = "serial0:115200n8";
+	};
+
+	aliases {
+		ethernet0 = &cp0_eth0;
+		ethernet1 = &cp0_eth2;
+	};
+
+	sfp_eth0: sfp-eth0 {
+		compatible = "sff,sfp";
+	};
+};
+
+&ap_sdhci0 {
+	status = "okay";
+};
+
+&cp0_i2c0 {
+	status = "okay";
+	clock-frequency = <100000>;
+};
+
+&cp0_spi0 {
+	status = "okay";
+};
+
+&cp0_spi1 {
+	status = "okay";
+};
+
+&cp0_pcie0 {
+	status = "okay";
+	phys = <&cp0_comphy0 0>;
+	phy-names = "pcie-phy0";
+};
+
+&cp0_pcie2 {
+	status = "okay";
+	phys = <&cp0_comphy5 2>;
+	phy-names = "pcie-phy0";
+};
+
+&cp0_sata0 {
+	status = "okay";
+
+	sata-port@0 {
+		status = "okay";
+		/* Generic PHY, providing serdes lanes */
+		phys = <&cp0_comphy1 0>;
+	};
+
+	sata-port@1 {
+		status = "okay";
+		/* Generic PHY, providing serdes lanes */
+		phys = <&cp0_comphy3 1>;
+	};
+};
+
+&cp0_usb3_1 {
+	status = "okay";
+};
+
+&cp0_mdio {
+	status = "okay";
+
+	cp0_phy0: ethernet-phy@0 {
+		reg = <0>;
+	};
+	cp0_phy1: ethernet-phy@3 {
+		reg = <1>;
+	};
+};
+
+&cp0_ethernet {
+	status = "okay";
+};
+
+&cp0_eth0 {
+	status = "okay";
+	/* Network PHY */
+	phy-mode = "10gbase-kr";
+	/* Generic PHY, providing serdes lanes */
+	phys = <&cp0_comphy2 0>;
+	managed = "in-band-status";
+	sfp = <&sfp_eth0>;
+};
+
+&cp0_eth2 {
+	status = "okay";
+	phy = <&cp0_phy1>;
+	phy-mode = "rgmii-id";
+};
diff --git a/arch/arm64/boot/dts/marvell/armada-7040-db-D.dts b/arch/arm64/boot/dts/marvell/armada-7040-db-D.dts
new file mode 100644
index 000000000000..b4d5696a50d5
--- /dev/null
+++ b/arch/arm64/boot/dts/marvell/armada-7040-db-D.dts
@@ -0,0 +1,126 @@
+// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
+/*
+ * Device Tree file for Marvell Armada 7040 development board
+ * This board file supports the D configuration of the board
+ */
+
+#include "armada-7040-db.dtsi"
+
+/ {
+	model = "Marvell Armada 7040 development board D setup";
+	compatible = "marvell,armada7040-db-D", "marvell,armada7040-db",
+		     "marvell,armada7040", "marvell,armada-ap806-quad",
+		     "marvell,armada-ap806";
+
+	chosen {
+		stdout-path = "serial0:115200n8";
+	};
+
+	aliases {
+		ethernet0 = &cp0_eth0;
+		ethernet1 = &cp0_eth2;
+	};
+
+	sfp_eth0: sfp-eth0 {
+		compatible = "sff,sfp";
+	};
+};
+
+&ap_sdhci0 {
+	status = "okay";
+};
+
+&cp0_pinctrl {
+	pca0_pins: pca0_pins {
+		marvell,pins = "mpp7";
+		marvell,function = "gpio";
+	};
+};
+
+&cp0_i2c0 {
+	status = "okay";
+	clock-frequency = <100000>;
+
+	expander0: pca9555@21 {
+		compatible = "nxp,pca9555";
+		pinctrl-0 = <&pca0_pins>;
+		/* GPIO 62 is defined as GPIO 30 of cp0_gpio1 */
+		interrupt-parent = <&cp0_gpio1>;
+		interrupts = <7 IRQ_TYPE_EDGE_FALLING>;
+		pinctrl-names = "default";
+		gpio-controller;
+		#gpio-cells = <2>;
+		reg = <0x21>;
+		/*
+		 * IO0_0: USB3_PWR_EN0	IO1_0: USB_3_1_Dev_Detect
+		 * IO0_1: USB3_PWR_EN1	IO1_1: USB2_1_current_limit
+		 * IO0_2: DDR3_4_Detect	IO1_2: Hcon_IO_RstN
+		 * IO0_3: USB2_DEVICE_DETECT
+		 * IO0_4: GPIO_0	IO1_4: SD_Status
+		 * IO0_5: GPIO_1	IO1_5: LDO_5V_Enable
+		 * IO0_6: IHB_5V_Enable	IO1_6: PWR_EN_eMMC
+		 * IO0_7:		IO1_7: SDIO_Vcntrl
+		 */
+	};
+};
+
+&cp0_nand {
+	status = "okay";
+};
+
+&cp0_pcie0 {
+	status = "okay";
+	phys = <&cp0_comphy0 0>;
+	phy-names = "pcie-phy0";
+};
+
+&cp0_pcie2 {
+	status = "okay";
+	phys = <&cp0_comphy5 2>;
+	phy-names = "pcie-phy0";
+};
+
+&cp0_sata0 {
+	status = "okay";
+
+	sata-port@1 {
+		status = "okay";
+		/* Generic PHY, providing serdes lanes */
+		phys = <&cp0_comphy3 1>;
+	};
+};
+
+&cp0_usb3_1 {
+	status = "okay";
+};
+
+&cp0_mdio {
+	status = "okay";
+
+	cp0_phy0: ethernet-phy@0 {
+		reg = <0>;
+	};
+	cp0_phy1: ethernet-phy@3 {
+		reg = <1>;
+	};
+};
+
+&cp0_ethernet {
+	status = "okay";
+};
+
+&cp0_eth0 {
+	status = "okay";
+	/* Network PHY */
+	phy-mode = "10gbase-kr";
+	/* Generic PHY, providing serdes lanes */
+	phys = <&cp0_comphy2 0>;
+	managed = "in-band-status";
+	sfp = <&sfp_eth0>;
+};
+
+&cp0_eth2 {
+	status = "okay";
+	phy = <&cp0_phy1>;
+	phy-mode = "rgmii-id";
+};
diff --git a/arch/arm64/boot/dts/marvell/armada-7040-db-E.dts b/arch/arm64/boot/dts/marvell/armada-7040-db-E.dts
new file mode 100644
index 000000000000..53e8ae09708f
--- /dev/null
+++ b/arch/arm64/boot/dts/marvell/armada-7040-db-E.dts
@@ -0,0 +1,100 @@
+// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
+/*
+ * Device Tree file for Marvell Armada 7040 development board
+ * This board file supports the E configuration of the board
+ */
+
+#include "armada-7040-db.dtsi"
+
+/ {
+	model = "Marvell Armada 7040 development board E setup";
+	compatible = "marvell,armada7040-db-E", "marvell,armada7040-db",
+		     "marvell,armada7040", "marvell,armada-ap806-quad",
+		     "marvell,armada-ap806";
+
+	chosen {
+		stdout-path = "serial0:115200n8";
+	};
+
+	aliases {
+		ethernet0 = &cp0_eth0;
+		ethernet1 = &cp0_eth1;
+	};
+};
+
+&cp0_i2c0 {
+	status = "okay";
+	clock-frequency = <100000>;
+};
+
+&cp0_i2c1 {
+	status = "okay";
+};
+
+&cp0_spi1 {
+	status = "okay";
+};
+
+&cp0_pcie0 {
+	status = "okay";
+	phys = <&cp0_comphy0 0>;
+	phy-names = "pcie-phy0";
+};
+
+&cp0_pcie2 {
+	status = "okay";
+	phys = <&cp0_comphy5 2>;
+	phy-names = "pcie-phy0";
+};
+
+&cp0_sata0 {
+	status = "okay";
+
+	sata-port@0 {
+		status = "okay";
+		/* Generic PHY, providing serdes lanes */
+		phys = <&cp0_comphy1 0>;
+	};
+
+	sata-port@1 {
+		status = "okay";
+		/* Generic PHY, providing serdes lanes */
+		phys = <&cp0_comphy3 1>;
+	};
+};
+
+&cp0_usb3_1 {
+	usb-phy = <&cp0_usb3_1_phy>;
+	status = "okay";
+};
+
+&cp0_mdio {
+	status = "okay";
+
+	phy0: ethernet-phy@0 {
+		reg = <0>;
+	};
+	phy1: ethernet-phy@2 {
+		reg = <1>;
+	};
+};
+
+&cp0_ethernet {
+	status = "okay";
+};
+
+&cp0_eth0 {
+	status = "okay";
+	/* Network PHY */
+	phy = <&phy0>;
+	phy-mode = "sgmii";
+	/* Generic PHY, providing serdes lanes */
+	phys = <&cp0_comphy2 0>;
+};
+
+&cp0_eth1 {
+	status = "okay";
+	phy = <&phy1>;
+	phy-mode = "rgmii-id";
+};
+
diff --git a/arch/arm64/boot/dts/marvell/armada-7040-db.dtsi b/arch/arm64/boot/dts/marvell/armada-7040-db.dtsi
new file mode 100644
index 000000000000..c50df47198f0
--- /dev/null
+++ b/arch/arm64/boot/dts/marvell/armada-7040-db.dtsi
@@ -0,0 +1,138 @@
+// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
+/*
+ * Device Tree file for Marvell Armada 7040 development board platform
+ */
+
+#include <dt-bindings/gpio/gpio.h>
+#include "armada-7040.dtsi"
+
+/ {
+	model = "Marvell Armada 7040 development board";
+	compatible = "marvell,armada7040-db", "marvell,armada7040",
+			"marvell,armada-ap806-quad", "marvell,armada-ap806";
+
+	memory@0 {
+		device_type = "memory";
+		reg = <0x0 0x0 0x0 0x80000000>;
+	};
+
+	cp0_reg_usb3_0_vbus: cp0-usb3-0-vbus {
+		compatible = "regulator-fixed";
+		regulator-name = "usb3h0-vbus";
+		regulator-min-microvolt = <5000000>;
+		regulator-max-microvolt = <5000000>;
+		enable-active-high;
+		gpio = <&expander0 0 GPIO_ACTIVE_HIGH>;
+	};
+
+	cp0_reg_usb3_1_vbus: cp0-usb3-1-vbus {
+		compatible = "regulator-fixed";
+		regulator-name = "usb3h1-vbus";
+		regulator-min-microvolt = <5000000>;
+		regulator-max-microvolt = <5000000>;
+		enable-active-high;
+		gpio = <&expander0 1 GPIO_ACTIVE_HIGH>;
+	};
+
+	cp0_usb3_0_phy: cp0-usb3-0-phy {
+		compatible = "usb-nop-xceiv";
+		vcc-supply = <&cp0_reg_usb3_0_vbus>;
+	};
+
+	cp0_usb3_1_phy: cp0-usb3-1-phy {
+		compatible = "usb-nop-xceiv";
+		vcc-supply = <&cp0_reg_usb3_1_vbus>;
+	};
+};
+
+&uart0 {
+	status = "okay";
+	pinctrl-0 = <&uart0_pins>;
+	pinctrl-names = "default";
+};
+
+&ap_sdhci0 {
+	bus-width = <4>;
+	no-1-8-v;
+	non-removable;
+};
+
+&cp0_i2c0 {
+	expander0: pca9555@21 {
+		compatible = "nxp,pca9555";
+		pinctrl-names = "default";
+		gpio-controller;
+		#gpio-cells = <2>;
+		reg = <0x21>;
+	};
+};
+
+&cp0_nand_controller {
+	nand-ecc-strength = <4>;
+	nand-ecc-step-size = <512>;
+	num-cs = <1>;
+	nand-on-flash-bbt;
+
+	status = "disabled";
+	partition@0 {
+		label = "U-Boot";
+		reg = <0 0x200000>;
+	};
+	partition@200000 {
+		label = "Linux";
+		reg = <0x200000 0xd00000>;
+	};
+	partition@1000000 {
+		label = "Filesystem";
+		reg = <0x1000000 0x3f000000>;
+	};
+};
+
+&cp0_spi1 {
+	spi-flash@0 {
+		#address-cells = <0x1>;
+		#size-cells = <0x1>;
+		compatible = "jedec,spi-nor";
+		reg = <0x0>;
+		spi-max-frequency = <20000000>;
+
+		partitions {
+			compatible = "fixed-partitions";
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			partition@0 {
+				label = "U-Boot";
+				reg = <0x0 0x400000>;
+			};
+
+			partition@400000 {
+				label = "Filesystem";
+				reg = <0x400000 0xc00000>;
+			};
+		};
+	};
+};
+
+&cp0_sdhci0 {
+	status = "okay";
+	bus-width = <4>;
+	no-1-8-v;
+	non-removable;
+};
+
+&cp0_usb3_0 {
+	usb-phy = <&cp0_usb3_0_phy>;
+	status = "disabled";
+	/* Generic PHY, providing serdes lanes */
+	phys = <&cp0_comphy1 0>;
+	phy-names = "usb";
+};
+
+&cp0_usb3_1 {
+	usb-phy = <&cp0_usb3_1_phy>;
+	status = "disabled";
+	/* Generic PHY, providing serdes lanes */
+	phys = <&cp0_comphy4 1>;
+	phy-names = "usb";
+};
-- 
2.17.1

